src/mame/misc/nabupc.cpp

```diff
@@ -0,0 +1,586 @@
+}
+
+// IRQ For HCCA Transmit
+WRITE_LINE_MEMBER(nabupc_state::hcca_tbre_w)
+{
+	BIT_SET(m_int_lines, 6, state);
+	update_irq();
+}
+
+// IRQ For Keyboard
+WRITE_LINE_MEMBER(nabupc_state::rxrdy_w)
+{
+	BIT_SET(m_int_lines, 5, state);
+	update_irq();
+}
+
+// IRQ For TMS9918a
+WRITE_LINE_MEMBER(nabupc_state::vdp_int_w)
+{
+	BIT_SET(m_int_lines, 4, state);
+	update_irq();
+}
+
+// IRQ For Option Slot 1
+WRITE_LINE_MEMBER(nabupc_state::j9_int_w)
+{
+	BIT_SET(m_int_lines, 3, !state);
+	update_irq();
+}
+
+// IRQ For Option Slot 2
+WRITE_LINE_MEMBER(nabupc_state::j10_int_w)
+{
+	BIT_SET(m_int_lines, 2, !state);
+	update_irq();
+}
+
+// IRQ For Option Slot 3
+WRITE_LINE_MEMBER(nabupc_state::j11_int_w)
+{
+	BIT_SET(m_int_lines, 1, !state);
+	update_irq();
+}
+
+// IRQ For Option Slot 4
+WRITE_LINE_MEMBER(nabupc_state::j12_int_w)
+{
+	BIT_SET(m_int_lines, 0, !state);
+	update_irq();
+}
+
+// IRQ Acknowledge
+IRQ_CALLBACK_MEMBER(nabupc_state::int_ack_cb)
+{
+	uint32_t vector = m_portb & 0xe;
+	return vector;
+}
+
+// Update CPU IRQ line
+void nabupc_state::update_irq()
+{
+	uint8_t interrupts = ~(m_porta & m_int_lines);
+	m_irq_in_prio = (m_irq_in_prio & 0xff00) | interrupts;
+
+	m_portb &= 0xf0;
+	uint8_t out = f9318(m_irq_in_prio);
+	m_portb |= bitswap(out, 2, 1, 0, 3);
+	m_maincpu->set_input_line(INPUT_LINE_IRQ0, !(BIT(out, 4)));
+}
+
+
+//**************************************************************************
+//  ADDRESS MAPS
+//**************************************************************************
+
+// RAM/ROM
+void nabupc_state::memory_map(address_map &map)
+{
+	map(0x0000, 0xffff).r(FUNC(nabupc_state::read_mem)).w(m_ram, FUNC(ram_device::write));
+}
+
+// IO Memory
+void nabupc_state::io_map(address_map &map)
+{
+	map.unmap_value_high();
+	map.global_mask(0xff);
+	map(0x00, 0x00).w(FUNC(nabupc_state::control_w));
+	map(0x40, 0x40).r(m_ay8910, FUNC(ay8910_device::data_r));
+	map(0x40, 0x41).w(m_ay8910, FUNC(ay8910_device::data_address_w));
+	map(0x80, 0x80).rw(m_hccauart, FUNC(ay31015_device::receive), FUNC(ay31015_device::transmit));
+	map(0x90, 0x91).rw(m_kbduart, FUNC(i8251_device::read), FUNC(i8251_device::write));
+	map(0xa0, 0xa1).rw(m_tms9928a, FUNC(tms9928a_device::read), FUNC(tms9928a_device::write));
+	map(0xb0, 0xb0).w("prndata", FUNC(output_latch_device::write));
+	map(0xc0, 0xcf).rw(m_bus, FUNC(bus::nabupc::option_bus_device::read<0>), FUNC(bus::nabupc::option_bus_device::write<0>));
+	map(0xd0, 0xdf).rw(m_bus, FUNC(bus::nabupc::option_bus_device::read<1>), FUNC(bus::nabupc::option_bus_device::write<1>));
+	map(0xe0, 0xef).rw(m_bus, FUNC(bus::nabupc::option_bus_device::read<2>), FUNC(bus::nabupc::option_bus_device::write<2>));
+	map(0xf0, 0xff).rw(m_bus, FUNC(bus::nabupc::option_bus_device::read<3>), FUNC(bus::nabupc::option_bus_device::write<3>));
+}
+
+uint8_t nabupc_state::read_mem(offs_t offset)
+{
+	if (offset < m_bios_size && (m_control & 1) == 0) {
+		return m_rom_base[offset];
+	}
+	return m_ram->read(offset);
+}
```

Comment:
So would the view allow writes to fall through to the underlying ram? That is if the ROM view is enabled and the machine tries to write to an address within the view would that write go to the RAM that the view is currently hiding? This is necessary for this machine and one of the reasons I ended up doing things this way, since I wasn't sure how to get this working with a declarative approach.

Reply:
> So would the view allow writes to fall through to the underlying ram? That is if the ROM view is enabled and the machine tries to write to an address within the view would that write go to the RAM that the view is currently hiding?

Yes, when you enable a view entry, any range with nothing mapped in it falls through to whatever’s underneath.  These kinds of ROM overlays are one of the most common use cases for it.

Reply:
Ok one more question regarding this, How am I supposed to check what the bios size is during machine_start like you suggested? Apparently input ports cannot be read during init, so i can't check to see what bios size has been selected for in machine_start.

Reply:
I've updated this to use a memory view overlay, but since I don't see anyway to read the config from within machine_start I'm doing that and installing the rom_entry during machine_reset at the moment.